packages/angular_devkit/core/src/workspace/core_spec.ts

```diff
@@ -0,0 +1,370 @@
+        }
+
+        return false;
+      },
+    };
+
+    try {
+      await readWorkspace(requestedPath, host);
+      fail();
+    } catch (e) {
+      expect(e.message).toContain('Unable to determine format for workspace path');
+    }
+
+    done();
+  });
+
+  it('errors when reading from specified file path with invalid specified format', async (done) => {
+    const requestedPath = '/path/to/workspace/angular.json';
+
+    const host: WorkspaceHost = {
+      async readFile(path) {
+        expect(path).toBe(requestedPath);
+
+        return '{ "version": 1 }';
+      },
+      async writeFile() {},
+      async isFile(path) {
+        return path === requestedPath;
+      },
+      async isDirectory(path) {
+        if (path !== requestedPath) {
+          fail();
+        }
+
+        return false;
+      },
+    };
+
+    try {
+      await readWorkspace(requestedPath, host, 12 as WorkspaceFormat);
+      fail();
+    } catch (e) {
+      expect(e.message).toContain('Unsupported workspace format');
+    }
+
+    done();
+  });
+
+  it('attempts to find/read from directory path', async (done) => {
+    const requestedPath = getSystemPath(normalize('/path/to/workspace'));
+    const expectedFile = getSystemPath(join(normalize(requestedPath), '.angular.json'));
+
+    const isFileChecks: string[] = [];
+    const host: WorkspaceHost = {
+      async readFile(path) {
+        expect(path).not.toBe(requestedPath);
+        expect(path).toBe(expectedFile);
+
+        return '{ "version": 1 }';
+      },
+      async writeFile() {},
+      async isFile(path) {
+        isFileChecks.push(path);
+
+        return path === expectedFile;
+      },
+      async isDirectory(path) {
+        if (path === requestedPath) {
+          return true;
+        }
+
+        fail();
+
+        return false;
+      },
+    };
+
+    await readWorkspace(requestedPath, host);
+    isFileChecks.sort();
+    expect(isFileChecks).toEqual([
+      getSystemPath(join(normalize(requestedPath), 'angular.json')),
+      getSystemPath(join(normalize(requestedPath), '.angular.json')),
+    ].sort());
+
+    done();
+  });
+
+  it('attempts to find/read only files for specified format from directory path', async (done) => {
+    const requestedPath = '/path/to/workspace';
+
+    const isFileChecks: string[] = [];
+    const host: WorkspaceHost = {
+      async readFile(path) {
+        expect(path).not.toBe(requestedPath);
+
+        return '{ "version": 1 }';
+      },
+      async writeFile() {},
+      async isFile(path) {
+        isFileChecks.push(path);
+
+        return true;
+      },
+      async isDirectory(path) {
+        if (path === requestedPath) {
+          return true;
+        }
+
+        fail();
+
+        return false;
+      },
+    };
+
+    await readWorkspace(requestedPath, host, WorkspaceFormat.JSON);
+
+    isFileChecks.sort();
+    expect(isFileChecks).toEqual([
+      getSystemPath(join(normalize(requestedPath), 'angular.json')),
```

Comment:
How do you verify other formats are not attempted? I think you need to not return file contents and verify all attempted file checks.

Reply:
Updated with a test only hook for the supported files.